### ISSUE TITLE
clip.py 1/2 speed of c++ code

### ISSUE THREAD

**Issue by nbecker** [serge-sans-paille/pythran#433]
I've updated to today's git and ran benchmark of clip.py vs. c++ code.  I got:
numpy << using numpy
2.81051397324
pythran non-vector << this is clip.py
0.938794851303
pythran vector << this is a different pythran version using masks (ignore for now)
1.63022494316
c++ << c++ version
0.424335002899

clip.py:
```python
#pythran export clip(complex128[], float64)
#pythran export clip(complex128[::], float64)
def clip (z, _max):
    out = np.empty (z.shape, dtype=z.dtype)
    for i in range (len(z)):
        if abs(z[i]) > _max:
            out[i] = z[i]/abs(z[i]) * _max
#            out[i] = limit1 (z[i]) * _max
        else:
            out[i] = z[i]

    return out
```
c++ version (inner part of code)
```c++
template<typename T>
inline T clip1 (T x, typename decomplexify<T>::type sat = typename decomplexify<T>::type(1)) {
  if (std::abs (x) > sat)
    return (x / std::abs(x)) * sat;
  else
    return x;
}
```

both compiled with -march=native -O3 -ffast-math

**Comment by pbrunet** [comment]
Hi neal,

I tried to use this implementation:
```Python
def clip(z, _max):
    return np.where(np.abs(z) > _max, z/np.abs(z) * _max, z)
```
I think it compute exactly the same thing.

You may notice that we have a bug to be fix with abs so we have to use np.abs in this case.

With this version, I have
python -m timeit "import numpy; a = numpy.arange(100000, dtype=complex); import test2" "test2.clip(a, 50000)"
100 loops, best of 3: 2.12 msec per loop

While with your version, I have:
python -m timeit "import numpy; a = numpy.arange(100000, dtype=complex); import test2" "test2.clip2(a, 50000)"
100 loops, best of 3: 1.82 msec per loop

Ho, we are slower using np.where?
Ok, here is a new version:
```Python
def clip(z, _max):
    b = np.abs(z) 
    return np.where(b > _max, z/b * _max, z)
```
python -m timeit "import numpy; a = numpy.arange(100000, dtype=complex); import test2" "test2.clip(a, 50000)"
1000 loops, best of 3: 1.94 msec per loop

It looks like the backend compiler can factorize common expression while it can't with numpy.where... Recomputing abs values is really expensive at least for complex numbers.

Finally, I compile it with the same options than you (previously, it was standard configurations: O2)
where code with manual CSA: 1000 loops, best of 3: 1.59 msec per loop
neal code: 1000 loops, best of 3: 1.56 msec per loop

with SIMD:
where code with manual CSA: 1.59 msec per loop
neal code : 1.56 msec per loop

with SIMD + OpenMP:
where code with manual CSA : 635 usec per loop
neal code : 1.2 msec per loop
neal code with explicit pragma omp on for loop : 644 usec per loop

Sorry, I can't compare with the c++ version but you can do it if you want. Also, what is the size of you inputs?

**Comment by nbecker** [comment]
with your version for pythran vector:
```python
#pythran export clip(complex128[], float64)
#pythran export clip(complex128[::], float64)
def clip(z, _max):
    b = np.abs(z) 
    return np.where(b > _max, z/b * _max, z)
```

I get:

numpy
2.94457507133
pythran non-vector
0.938816070557
pythran vector
1.0420460701
c++
0.424599885941

I'm testing with length 10^6 vector (contiguous) input

What is 'with SIMD'?

**Comment by pbrunet** [comment]
It is an experimental feature in pythran to use AVX and SSE vectors for computation (thanks to boost.simd). It can be enable with -DUSE_BOST_SIMD but np.where is marked as not vectorizable so I think it is the reason why it doesn't work.

For openmp, you just have to add the -fopenmp flags on the commande line.

for you c++ code, I guess you are using a raw array?
```C++
std::complex<double>* z = new std::complex<double>[1000000];
```
(I try to understand if the cost is coming from numpy.array index computation in pythran or if it is the array creation which is costly).

**Comment by jfalcou** [comment]
Le 17-04-2015 15:50, pbrunet a écrit :
> It is an experimental feature in pythran to use AVX and SSE vectors
> for computation (thanks to boost.simd). It can be enable with
> -DUSE_BOST_SIMD but np.where is marked as not vectorizable so I think
> it is the reason why it doesn't work.

is np.where basicaly cond ? a : b   ?

If yes, boost::simd provides it as simd::if_else(cond,a,b);

**Comment by pbrunet** [comment]
It is exactly this pattern. I can't fix it for now but thanks for the tip jfalcou

**Comment by nbecker** [comment]
```c++
template<typename T>
inline T clip1 (T x, typename decomplexify<T>::type sat = typename decomplexify<T>::type(1)) {
  if (std::abs (x) > sat)
    return (x / std::abs(x)) * sat;
  else
    return x;
}

template<typename out_cont_t, typename in_cont_t, typename T>
inline out_cont_t clip (const in_cont_t& in, T sat) {
  out_cont_t out (boost::size (in));
  typename boost::range_const_iterator<in_cont_t>::type i = boost::begin (in);
  typename boost::range_const_iterator<in_cont_t>::type e = boost::end (in);
  typename boost::range_iterator<out_cont_t>::type o = boost::begin (out);
  for (; i != e; ++i, ++o)
    *o = clip1<typename boost::range_value<in_cont_t>::type> (*i, sat);

  return out;
}
...
  def ("clip", &clip<nd::Array<Complex,1,1>, nd::Array<Complex,1,0>, double>, (arg("in"), arg ("sat")=1));

```

This is using the ndarray library from https://github.com/ndarray/ndarray and boost::python

**Comment by serge-sans-paille** [comment]
@nbecker sorry foir resurrecting the thread, but it seems to me the situation has improved (running the test locally with master version). Can you confirm we know match c++ performance?

**Comment by nbecker** [comment]
python test_clip.py
numpy
6.43687415123
pythran non-vector
2.25750184059
pythran vector
2.25895500183
c++
1.85485982895
---------------
pythran non-vector version is:
#pythran export clip(complex128[], float64)
#pythran export clip(complex128[::], float64)
def clip (z, _max):
    out = np.empty (z.shape, dtype=z.dtype)
    for i in range (len(z)):
        if abs(z[i]) > _max:
            out[i] = z[i]/abs(z[i]) * _max
#            out[i] = limit1 (z[i]) * _max
        else:
            out[i] = z[i]

    return out

pythran vector version is:
#pythran export clip(complex128[], float64)
#pythran export clip(complex128[::], float64)
def clip(z, _max):
    b = np.abs(z)
    return np.where(b > _max, z/b * _max, z)


On Tue, May 31, 2016 at 12:55 PM, serge-sans-paille <
notifications@github.com> wrote:

> @nbecker <https://github.com/nbecker> sorry foir resurrecting the thread,
> but it seems to me the situation has improved (running the test locally
> with master version). Can you confirm we know match c++ performance?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serge-sans-paille/pythran/issues/433#issuecomment-222750898>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHK0OQFBMCLpqxxEBKRRBYsWAiyvDvNks5qHGgCgaJpZM4ECtY6>
> .
>



-- 
*Those who don't understand recursion are doomed to repeat it*

**Comment by serge-sans-paille** [comment]
@nbecker can you try with
```
[pythran]
complex_hook = True
```

in your ``~/.pythranrc`` ?
To be sure : the c++ code uses double and not float ? you're compiling it with the same gcc, optimization level -O2 ?

**Comment by serge-sans-paille** [comment]
@nbecker got it! We're using nt2 version of abs, and it appears to introduce a large overhead... falling back to std::abs fixes the issue cc @jfalcou .

**Comment by jfalcou** [comment]
is it scalar or simd abs ? For which type ?

**Comment by serge-sans-paille** [comment]
scalar abs, for complex numbers

**Comment by serge-sans-paille** [comment]
(from the nt2 version frozen by pythran)

**Comment by jfalcou** [comment]
you may want to compile with -DBOOST_SIMD_NO_INVALIDS to remove all the corner case handling of inf, nan etc .

See the macro there : 
https://github.com/jfalcou/nt2/blob/master/modules/boost/simd/sdk/include/boost/simd/sdk/config.hpp

**Comment by serge-sans-paille** [comment]
@jfalcou that's not enough (it does not change anything...). i see an unexpcted call to ilogb in the resulting binary...

**Comment by jfalcou** [comment]
ilogb wow. 
abs on complex is hypot(real,imag) and hypot use the bit perfect algo based on ldexp/frexp.
Care to fill an issue on NT2 about this and assign jtlap ?

**Comment by serge-sans-paille** [comment]
@jfalcou see https://github.com/NumScale/boost.simd/issues/147

**Comment by jfalcou** [comment]
I was speakign of the NT2 repo ;)

**Comment by serge-sans-paille** [comment]
#574 should fix the issue

**Comment by serge-sans-paille** [comment]
@nbecker the fix is in master, if your tests are ok, I'll close the issue.

**Comment by nbecker** [comment]
python test_clip.py
numpy
6.48566198349
pythran non-vector
2.256513834
pythran vector
1.27110791206
c++
1.80149412155

looks good!

On Fri, Jun 3, 2016 at 8:03 AM, serge-sans-paille <notifications@github.com>
wrote:

> @nbecker <https://github.com/nbecker> the fix is in master, if your tests
> are ok, I'll close the issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serge-sans-paille/pythran/issues/433#issuecomment-223561462>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHK0DX_1uLOTrGqh4krh8ijIWBdHb3_ks5qIBgpgaJpZM4ECtY6>
> .
>



-- 
*Those who don't understand recursion are doomed to repeat it*

**Comment by serge-sans-paille** [comment]
@nbecker and if you compile with pythran -03 ?

**Comment by nbecker** [comment]
python test_clip.py
numpy
6.52374792099
pythran non-vector
2.1869969368
pythran vector
1.35169816017
c++
1.85517692566

But I already had this in .pythranrc:
cflags=-std=c++11 -fno-math-errno -w -O2 -ffast-math -march=native

On Fri, Jun 3, 2016 at 8:17 AM, serge-sans-paille <notifications@github.com>
wrote:

> @nbecker <https://github.com/nbecker> and if you compile with pythran -03
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serge-sans-paille/pythran/issues/433#issuecomment-223563834>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHK0J9tphZD9eSEOlWNPv6lRUqM0Ic_ks5qIBtVgaJpZM4ECtY6>
> .
>



-- 
*Those who don't understand recursion are doomed to repeat it*

**Comment by serge-sans-paille** [comment]
ok, that's enough for me, closing the issue. Nice to see the vector version runs faster than the C++ one ;-) removing ffast-math would be less unfair though

**Comment by nbecker** [comment]
I believe the c++ version used the same options, -ffast-math -march=native

On Fri, Jun 3, 2016 at 8:21 AM, serge-sans-paille <notifications@github.com>
wrote:

> Closed #433 <https://github.com/serge-sans-paille/pythran/issues/433>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serge-sans-paille/pythran/issues/433#event-681061108>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHK0MbNcdB0px6BCJAKH4nriMlFfhaRks5qIBxcgaJpZM4ECtY6>
> .
>



-- 
*Those who don't understand recursion are doomed to repeat it*
